### PR TITLE
Feature/cache short second attempt

### PR DESCRIPTION
**Pull request by lasote**
Changelog:  omit
Docs: omit

- To discuss the length we want. 
- This alleviates the PATH env variables limit and windows users are forced to use the user home folder instead of `C:/`
- Introduced assert and simplified apparently dead code

### CODE REVIEW

File: conan/cache/cache.py [hunk 1]
```diff
@@ -53,7 +54,9 @@ def _get_tmp_path():
 
     @staticmethod
     def _get_path(ref: ConanReference):
-        return md5(ref.full_reference)
+        value = ref.full_reference.encode("utf-8")
```
Comment (memsharded) [inline conan/cache/cache.py]: can't we assume at this point that all Conan strings are encoded to utf-8?

**Review by memsharded — APPROVED** [review]
There is pending an improvement in the graph management in the "install binaries" step, that would mitigate this ugly pattern, but that should be done later.

File: conan/cache/cache.py [hunk 2]
```diff
@@ -53,7 +54,9 @@ def _get_tmp_path():
...
+        sha = sha256(value)
```
Comment (memsharded) [inline conan/cache/cache.py]: BTW, a common trick to reduce from 2->1,3 ratio characters is to encode the hash from hexadecimal to base64

File: conan/cache/cache.py (hunk 1 above)
Comment (lasote) [reply conan/cache/cache.py]: at that point, according to test failures, I would say it is an unicode object

File: conan/cache/cache.py (hunk 2 above)
Comment (lasote) [reply conan/cache/cache.py]: Nice idea

File: conan/cache/cache.py (hunk 1 above)
Comment (memsharded) [reply conan/cache/cache.py]: Oh, I see, the thing is that ``sha256`` requires bytes, didn't realize it.

**Comment by czoido** [comment]
What about making this configurable (per package or global) and by default using the whole path? Maybe for most of the cases we don't need the shortening and it reduces the probability of collision?

Also, and out of the scope of this PR but related, regarding the length of the environment variables, maybe we could also shorten a little bit more when generating the environment scripts. I mean that for example we generate a _conanrunenv.sh_ like this:

```
echo Configuring environment variables

export PATH="/path/to/conan/home/that/may/be/long/d8eea279520277c9a1cd8a3d91e45b79/package/bin:$PATH"
```
 but could that also be shortened using this?

```
echo Configuring environment variables
export CONAN_PATH = "/path/to/conan/home/that/may/be/long"
export PATH="$CONAN_PATH/d8eea279520277c9a1cd8a3d91e45b79/package/bin:$PATH"
```

In case you have many transitive dependencies you could also save lots of characters from the max of the PATH environment variable. Don't know if this have been discussed before or not...

**Comment by memsharded** [comment]
> What about making this configurable (per package or global) and by default using the whole path? Maybe for most of the cases we don't need the shortening and it reduces the probability of collision?

I don't think so at this stage, let's keep it as simple as possible.

I agree with the environment improvements, but that should be a different effort.

**Comment by czoido** [comment]
I'm not convinced about shortening the paths by default. We have two risks here, in one hand we have the probability of someone having problems in Windows with the paths being too long and in the other hand someone having lots of packages that will have collisions in the cache. With the time we know that both of them will happen. I don't know which one of those is more likely...

Adding some numbers here so we can evaluate the risk of collision vs the risk of having problems with path length in Windows:

| chars for hash (not base64) | number recipes + binaries | % chance of collision |
|-----------------------------|---------------------------|-------------|
| 8                           | 1000                      | 0.011%      |
| 8                           | 10.000                    | 1.15%       |
| 8                           | 100.000                   | 68.78%      |
| 10                          | 1000                      | 4.56*10^-5  |
| 10                          | 10.000                    | 0.004%      |
| 10                          | 100.000                   | 0.45%       |
| 12                          | 1000                      | 1.7*10^-7   |
| 12                          | 10.000                    | 1.7*10^-5   |
| 12                          | 100.000                   | 0.0017      |
| 16                          | 1000                      | 2.7*10^-12  |
| 16                          | 10.000                    | 2.7*10^-10  |
| 16                          | 100.000                   | 2.7*10^-8   |
| 16                          | 1.000.0000                | 2.7*10^-6   |

Used
```
chance_of_collision = 1 - e^(-n^2 / (2 * d))
 d: 2^(bits_used_for_hash)
 bits_used_for_hash: 4*truncated_chars_length
 n: number_packages, number of hashes to check the probability for
```
to calculate. More info: https://stackoverflow.com/questions/51622061/how-can-i-calculate-the-impact-on-collision-probability-when-truncating-a-hash and https://preshing.com/20110504/hash-collision-probabilities/

Maybe at least we should start with a higher number of bytes and see if someone has problems with collisions or path length. With 16 characters (can be reduced with base64) it should be pretty safe.

**Comment by lasote** [comment]
I've removed the base64 trick because:
a) "/" is a valid character and it mess with the paths 
b) The casing "a" vs "A" can cause additional collisions in Windows/Macos

**Comment by lasote** [comment]
@memsharded check if we should increment to 16

**Comment by memsharded** [comment]
> I've removed the base64 trick because:

You applied it in the wrong place. You can take the base64 of the first N chars of the hexadecimal sha256. That will not have any problem, and will be faster to compute.

**Comment by memsharded** [comment]
> @memsharded check if we should increment to 16

Yes, I'd say we increment to 12 base64 characters => first 16 hexadecimal chars of the sha256

**Comment by lasote** [comment]
> You applied it in the wrong place. You can take the base64 of the first N chars of the hexadecimal sha256. That will not have any problem, and will be faster to compute.

I don't get why it won't have any problem. The output of a base64 contains "/" and letters with casing "a", "A".

**Comment by memsharded** [comment]
You are right, I didn't realize that base64 includes the backslash
character, and indeed the casing could cause issues.

Let's keep the hexadecimal then

**Review by czoido — APPROVED** [review]

**Pull request merged**